cpp/src/arrow/ipc/adapter.cc

```diff
@@ -243,6 +271,26 @@ class RowBatchReader::Impl {
       return MakePrimitiveArray(
           type, field_meta.length, data, field_meta.null_count, null_bitmap, out);
     }
+    if (IsListType(type.get())) {
+      std::shared_ptr<Buffer> offsets;
+      if (field_meta.length > 0) {
+        RETURN_NOT_OK(GetBuffer(buffer_index_++, &offsets));
+      } else {
+        offsets.reset(new Buffer(nullptr, 0));
```

Comment:
After further consideration I think we should always serialize the offsets. I don't think it is worth consuming code have to special case for 0 length lists.